### DISCUSSION TITLE
ViewState and ViewAction

### DISCUSSION THREAD

**Discussion by ibrahimkteish** [pointfreeco/isowords#152] [Q&A] [answered]
Hello,

Thank you for this amazing project, I am trying to replicate your work in my personal project.

I just have a question regarding `ViewState` and `ViewAction` inside a `SwiftUI View` sometimes you add them sometime you don't,  in some `Views` you directly create  a `ViewStore` from the injected `store` in `init`, so my question is based on what you decide to create `ViewState` and `ViewAction`

**Comment by stephencelis** [comment, marked as answer]
@ibrahimkteish Scoping a view store's state (_e.g._, to `ViewState`) is important to avoid recomputing the view hierarchy too often. SwiftUI observes the view store as an `ObservableObject` based on its state. If you don't scope the view store to just the state a given view needs, performance may suffer, especially in the outer layers of your app, where more global state is available to your store. In isowords you'll see that most of our scoping to `ViewState` occurs in the outer layers of the app. At leaf views that don't pass the store any further, we are less apt to need to scope.

We talked about it most recently in this episode: https://www.pointfree.co/episodes/ep151-composable-architecture-performance-view-stores-and-scoping

And we developed the concept here: https://www.pointfree.co/episodes/ep95-adaptive-state-management-state

`ViewAction`, meanwhile, is more of a way of giving the view less control (hiding actions that should only be sent from effects) and adapting a reducer for many platforms where views may have different action names depending on the platform. We first developed the concept here:

https://www.pointfree.co/episodes/ep96-adaptive-state-management-actions

**Reply by ibrahimkteish** [reply]
Thank you so much for the explanation :)
